auth/http/http.go

```diff
@@ -0,0 +1,189 @@
+package auth
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"io"
+	"net/http"
+	"net/url"
+	"strconv"
+
+	mqtt "github.com/mochi-mqtt/server/v2"
+	"github.com/mochi-mqtt/server/v2/packets"
+)
+
+// Hook is a hook that makes http requests to an external service
+type Hook struct {
+	httpclient     *http.Client
+	aclhost        *url.URL
+	clientauthhost *url.URL
+	superuserhost  *url.URL // currently unused
+	mqtt.HookBase
+}
+
+// Options is a struct that contains all the information required to configure the http hook
+// It is the responsibility of the configurer to pass a properly configured RoundTripper that takes
+// care other requirements such as authentication, timeouts, retries, etc
+type Options struct {
+	ACLHost                  *url.URL
+	SuperUserHost            *url.URL
+	ClientAuthenticationHost *url.URL // currently unused
+	RoundTripper             http.RoundTripper
+}
+
+// ClientCheckPOST is the struct that is sent to the client authentication endpoint
+type ClientCheckPOST struct {
+	ClientID string `json:"clientid"`
+	Password string `json:"password"`
+	Username string `json:"username"`
+}
+
+// ACLCheckPOST is the struct that is sent to the acl endpoint
+type ACLCheckPOST struct {
+	Username string `json:"username"`
+	ClientID string `json:"clientid"`
+	Topic    string `json:"topic"`
+	ACC      string `json:"acc"`
+}
+
+// ID returns the ID of the hook
+func (h *Hook) ID() string {
+	return "http-auth-hook"
+}
+
+// Provides returns whether or not the hook provides the given hook
+func (h *Hook) Provides(b byte) bool {
+	return bytes.Contains([]byte{
+		mqtt.OnACLCheck,
+		mqtt.OnConnectAuthenticate,
+	}, []byte{b})
+}
+
+// Init initializes the hook with the given config
+func (h *Hook) Init(config any) error {
+	if config == nil {
+		return errors.New("nil config")
+	}
+
+	authHookConfig, ok := config.(Options)
+	if !ok {
+		return errors.New("improper config")
+	}
+
+	if !validateConfig(authHookConfig) {
+		return errors.New("hostname configs failed validation")
+	}
+
+	h.httpclient = NewTransport(authHookConfig.RoundTripper)
+
+	h.aclhost = authHookConfig.ACLHost
+	h.clientauthhost = authHookConfig.ClientAuthenticationHost
+	h.superuserhost = authHookConfig.SuperUserHost
+	return nil
+}
+
+// OnConnectAuthenticate is called when a client attempts to connect to the server
+func (h *Hook) OnConnectAuthenticate(cl *mqtt.Client, pk packets.Packet) bool {
+
+	payload := ClientCheckPOST{
+		ClientID: cl.ID,
+		Password: string(pk.Connect.Password),
+		Username: string(pk.Connect.Username),
+	}
+
+	resp, err := h.makeRequest(http.MethodPost, h.clientauthhost, payload)
+	if err != nil {
+		h.Log.Error("error occurred while making http request", "error", err)
+		return false
+	}
+
+	// Block on proper 4xx response
+	if resp.StatusCode == http.StatusUnauthorized || resp.StatusCode == http.StatusForbidden {
```

Comment:
It might be extra flexible to provide a callback function here to validate the `resp` and use this as a sensible default (maybe it returns bool).